### ISSUE TITLE
Coverage mechanism vs input feeding

### ISSUE THREAD

**Issue by atsb1372** [abisee/pointer-generator#47]
Hello,
My question is from the paper than the code. So can I conclude that the coverage mechanism is an alternative to input feeding approach in [Effective Approaches to Attention-based Neural Machine Translation](https://arxiv.org/pdf/1508.04025.pdf)? Since we are using both of them to have some kind of memory of past alignments and avoid repetition? Please correct me if this is wrong. Thanks

**Comment by shahbazsyed** [comment]
@atsb1372 From what I understand, you can think of the coverage mechanism as more of an improvisation/boost to the attention mechanism. It allows the attention mechanism to remember what words it has already covered, and avoids attending to them again and again once they have been attended to initially.  This is done by maintaining a __coverage vector__ and the __attention distribution__ , and any overlap between them is penalized. The coverage vector is just the sum of all the attention distributions so far, which is calculated at each step of the decoder. This allows the attention mechanism in the next step to see if the word it is about to consider now, has already been attended previously.

The input feeding approach on the other hand does not take any measures to tell the decoder that a certain word was __already__ attended a while ago. It just informs what words were considered for alignment in the past which can include an already covered word if the attention mechanism considers it again and again, which caused the __repeating words__ problem in the first place. 

Correct me if I'm wrong!
